### DISCUSSION TITLE
Call Program not working as expected

### DISCUSSION THREAD

**Discussion by Robjones6146** [artisan-roaster-scope/artisan#1285] [Q&A]
**Call Program won't run this**    C:/Program Files/notepad.exe  or  C:/Windows/notepad.exe
Any ideas Please?

I would like to open Notepad, create a file called text.txt or test.csv, and then write some text to the file and save it.

Does anybody know how to do this?
Thank you for your help.

Artisan 2.8.5 (ed1bfa9)

**Comment by Robjones6146** [comment]
Call Program won't run this C:/Program Files/notepad.exe or C:/Windows/notepad.exe

Is this normal behavior not to run this program?

**Comment by roasterdave** [comment]
The Call Program command is intended to run programs/scripts that require no human interaction.  It is not intended to open a windowed application program like Notepad.  On Windows, Call Program starts programs in a windowless mode.   Notepad is indeed started by Call Program but it is hidden from view.  Check your task manager for these windowless Notepad processes.  

Good news -- there is a way to open Notepad from Call Program.  Create a one line batch file that uses the "start" command to open Notepad.exe.  
`start c:\windows\notepad.exe`

Now set Call Program to start the batch file.  The batch file will open Notepad in a window and then the batch file exits.

**Comment by Robjones6146** [comment]
**Thanks for your help and advice.**

I don't need the window to open. (I thought it had to) I took a look at the Task Manager and indeed lots and lots of entries!

I want to write to test.csv or .txt file five times during the roast. I want to record the Start time, dry end time, FC time, Drop time and Cool end time.

Possibly I might add temperature if I can access Artisan variables.

I will then import this information into my Filemaker Pro application at the end of each roast.

I have tried a few batch scripts, but none have worked.
I will use an If Alarm on START to record the time
I want to write the current time on the first line, then an alarm at DRY END to record the current time on a new line or comma-separated ...etc.
I also want to overwrite the file on each new roast.

I assume I start with C:/Windows/notepad.exe
How do I create, write, and update the file?

**Comment by Robjones6146** [comment]
**Thanks Dave. Made some progress.**
Buttons >> Call Program >> C:/Users/Robin/Documents/batch.bat

**Batch file:**
@echo off
echo %date% %time% >> C:/Users/Robin/Documents/batchtest.txt
exit

**Output:**
26/10/2023  6:59:11.75
26/10/2023  7:02:21.57 (next on new line!)

**Now all I need is temperatures.**
Is there a way to access variables in Artisan that I can write to this file?

**Reply by poundy** [reply]
I can't add anything about accessing BT and/or ET values as script parameters to your batch job. 

TBH, I sort of understand what you're trying to do, but I think I would approach this differently. 

If what you want is a SUMMARY FILE that you can load into your databse, then why aren't you considering your ALOG files as INPUT and then write a parser that pulls the info you want our ? They have everything recorded for a roast, are a single file for each roast, and you can do this between batches or at the end of the roast session and still be "complete". 

The batch file you have written is very simple - to be useful I suspect you need it be smarter than it is now, for example if you start a new batch your existing batchtest.txt file will just contain more info and not report whether it's a charge, FCS, or drop time...  and this certainly isn't what I expect the call program method was intended to do, which is why I suggest rethinking what you want to achieve (which we really don't know, just the bits you've mentioned along the way).

**Comment by Robjones6146** [comment]
**Thanks for your input.** Thats a good idea.

I already parse the .alog file. I create a background roast profile for each bean/roast. I then run a script in Filemaker that makes a few changes and then saves the file as generic.alog
The roaster clicks the **Load FM Roast button** in Artisan and the background roast profile is loaded. 

**Some context**
This avoids any mistakes. There are about 50 different green beans and over 130 roast profiles, all roasted to order, from 250g bag to 400 times 1kg bags. About 40 - 80 roasts per day from 2kg to 30kg batches.
Filemaker imports orders every 2 hours (from Shopify > Linnworks > Filemaker), then collates and optimizes all the bean/roast and order sizes into roast orders across four different size drum roasters (2kg to 30kg). Labels and Pickslips are printed every 30 minutes for completed bean/roasts. This is a very simplified version of what Filemaker does.

The four roasters will be replaced by one fluid bed roaster controlled by Filemaker and Artisan. From weighing green beans in, loading, roasting, weighing roasted beans out, transporting, destoning, labels, barcodes, weighing and bagging, and extensive reporting.

I want to automate as much of the roast process as possible. I want consistency and I also want to avoid human error. 

I did consider the .alog file, but I want to automate the process. How would I save the .alog file using an Alarm at the end of the roast?
Once I have the file, I can import it into Filemaker and get all the fields I want.

This is my current CSV file output (now appending to the same line with a comma)
**Batch file:**
@echo off
echo |set /p=%time%,<nul >> C:/Users/Robin/Documents/fmRoast.csv
exit
**Output file:**
8:21:08.04, 8:21:17.34, 8:21:21.29, 8:21:26.18, 8:21:30.08, (five times)

**Reply by roasterdave** [reply]
`How would I save the .alog file using an Alarm at the end of the roast?`
Use the Config>> Autosave feature.  No alarm necessary. Your profile is automatically saved at the end of each roast.  Click the Help button to see the enormous control you have for setting the file name.

Please consider Artisan's value to your roasting operation and make a [contribution to sustain Artisan](https://artisan-scope.org/donate/). Artisan is offered as open source however there are real costs to develop and support it. Artisan cannot exist without support from its users.  Thank you.

**Reply by poundy** [reply]
Yes, Autosave is what you need. i personally automate the OFF command after DROP by using a timer alarm, since at that point I'm hands off on the computer until I am ready to use ON again, at which time if the temperatures are above a minimum I set, I have an alarm to automate the START too - meaning I only need click DROP, ON, CHARGE to go from one batch to the next. So something like those are possibly helpful for your needs.

**Reply by Robjones6146** [reply]
**Thank you both.**
I do automate the START from Load Beans and OFF, then switch back on.
Autosave on!
I have found all the variable names I need to parse from the .alog file.

I am not the end user. The company will make a donation once I have the roaster up and running. The roaster is in a shipping container heading for the UK.
